urld829b80034_wjch111_PTSP2013Wang_tgz-pJ8-controllers_momcts_momctsCore_TestPlannerJ8/src/framework/graph/PathFinder.java

```diff
@@ -0,0 +1,382 @@
+                    // Cost from current to connected (edge cost)
+                    double dA1 = edg.cost();
+
+                    //Path to this node
+                    PathCH pc = new PathCH();
+                    pc.p = D1;
+                    pc.destID = connectedID;
+
+                    //If the new cost is smaller.
+                    double newCost = DA.m_cost + dA1;
+                    if(D1.m_cost > newCost)
+                    {
+                        //update cost
+                        Path newD1 = new Path(DA);
+                        newD1.m_destinationID = connectedID;
+                        newD1.m_cost += dA1;
+                        //update path
+                        pc.p = newD1;
+                        newD1.m_points.add(connectedID);
+                        setShortestPath(newD1);
+                    }
+
+                    //Set cost, used by priority queue to navigate more efficiently
+                    connected.m_g = pc.p.m_cost;
+                    connected.m_f = pc.heuristicCost = pc.p.m_cost + heuristic(connectedID, a_path.m_destinationID);
+                    if(!toEvaluateSetMarker.contains(connected.id()))
+                    {
+                        //Mark this node as evaluated.
+                        toEvaluateSet.add(connected);
+                        toEvaluateSetMarker.add(connected.id());
+                    }
+
+                }
+            }
+        }
+
+        //If we didn't find the destination, the path could not be extracted.
+        return false;
+    }
+
+    /**
+     * Heuristic for A*: euclidean distance.
+     * @param a_or origin node id.
+     * @param a_dest destination node id.
+     * @return the euclidean distance between origin and destination.
+     */
+    private double heuristic(int a_or, int a_dest)
+    {
+        return m_graph.getNodes().get(a_or).euclideanDistanceTo(m_graph.getNodes().get(a_dest));
+    }
+
+    /**
+     * Returns all the shortest paths in cache from a given node.
+     * @param a_origin origin node id.
+     * @return a TreeMap with all the paths from the origin node.
+     */
+    private TreeMap<Integer, Path> getShortestPaths(int a_origin){return m_shortestPaths.get(a_origin);}
+
+}
+
+/**
+ * CLASS to compare two nodes by the graph cost.
+ */
+class NodeComparatorG implements Comparator<Node>
+{
+    /**
+     * Compares two nodes.
+     * @param x one node.
+     * @param y another node.
+     * @return returns -1, 0, or 1 depending on the nodes: -1 if x has a smaller graph cost than y, 1 if opposite, 0 if it is the same cost.
+     */
+    public int compare(Node x, Node y)
+    {
+        // Assume neither PathCH is null. Real code should
+        // probably be more robust
+        if (x.m_g < y.m_g)
+        {
+            return -1;
+        }
+        if (x.m_g > y.m_g)
+        {
+            return 1;
+        }
+        return 0;
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Similar code fragments were detected in the same file at the following lines: 289:297, 317:325.  
Refactoring can help improve code maintainability. Consider reducing duplicate code by extracting it into a separate method. You can then replace duplicated code with calls to this new method.